iceoryx_utils/test/integrationtests/test_resizeable_lockfree_queue_stresstest.cpp

```diff
@@ -0,0 +1,829 @@
+// Copyright (c) 2020 Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "test.hpp"
+
+#include "iceoryx_utils/concurrent/resizeable_lockfree_queue.hpp"
+
+// Remark: It would be nice to have way to configure the (maximum) runtime in a general way.
+
+using namespace ::testing;
+
+#include <atomic>
+#include <list>
+#include <numeric>
+#include <random>
+#include <thread>
+#include <utility>
+#include <vector>
+
+namespace
+{
+struct Data
+{
+    Data(int id = 0, size_t count = 0)
+        : id(id)
+        , count(count)
+    {
+    }
+
+    int id{0};
+    size_t count{0};
+
+    void print()
+    {
+        std::stringstream s;
+        s << "data id " << id << " count " << count << std::endl;
+        std::cout << s.str();
+    }
+};
+
+template <typename Queue>
+void produce(Queue& queue, int id, int iterations)
+{
+    Data d(id, 0);
+    for (int i = 0; i < iterations; ++i)
+    {
+        d.count++;
+        while (!queue.tryPush(d))
+        {
+        }
+    }
+}
+
+template <typename Queue>
+void consume(Queue& queue, std::atomic<bool>& run, size_t expectedFinalCount, int maxId, bool& testResult)
+{
+    bool error = false;
+
+    std::vector<size_t> lastCount(maxId + 1, 0);
+
+    while (run || !queue.empty())
+    {
+        auto popped = queue.pop();
+        if (popped.has_value())
+        {
+            auto& value = popped.value();
+            if (lastCount[value.id] + 1 != value.count)
+            {
+                error = true;
+            }
+
+            lastCount[value.id] = value.count;
+        }
+    }
+
+    for (int i = 1; i <= maxId; ++i)
+    {
+        if (lastCount[i] != expectedFinalCount)
+        {
+            error = true;
+        }
+    }
+
+    testResult = !error;
+}
+
+/// remark: a possible rework could try to avoid storing the popped values for check with multiple consumers
+/// since this would allow us to run the test much longer (currently we will exhaust memory
+/// by using the list), but this rework is somewhat nontrivial
```

Comment:
I didn't skim the test, so please forgive me my ignorance, but how about the following proposal.
- given are four arrays with atomics `std::array<std::atomic<uint64_t, QueueCapacity>> queueOnePushCounter, queueOnePopCounter, queueTwoPushCounter, queueTwoPopCounter;`
- given is a full queue with indices from `0` to `QueueCapacity-1`, e.g. `queueOne` and an empty queue, e.g. `queueTwo`
- half of the threads pop from `QueueOne` and push to `QueueTwo`, the other half pop from `QueueTwo` and push to `QueueOne`
- after a pop and before a push, the corresponding counter is incremented (the index of the counter is the value of the pop operation)
- let this run for a day
- stop all threads and sanitize the counter by popping `QueueTwo` empty and pushing the indices to `QueueOne`
- `QueueOne` must be full again
- `QueueOne` must have all the indices from `0` to `QueueCapacity`
- the four counter must be the same at a given index

Reply:
Need to think about this, will add another test case and if it works may replace this one. These tests are kind of hard to understand and have increasingly complex setups, I know. Partly because of the multi threaded nature, partly because of the fact that the tested invariants themselves are fairly complex.

Reply:
I do not fully understand, but how this check for order? Currently the tests that uses this has multiple producers and 2 consumers.

The test verifies that no data is lost and also that it arrives in order. I.e. if a consumer sees a value X with counter c from producer i, the next value from producer X it sees will hve at least counter c+1, but possibly more because the other consumer might consumed it. This ordering is easy to check though without unbounded additional memory.

I think one queue is enough, the producers just have to produce modulo capacity and the consumers increment the respective counter. Yes, I see a possible improvement now.

Reply:
Reworked those tests considerably, checking completeness with an apporach similar to the one you proposed. Tests got a bit easier to understand I think but are still complex (unavoidable).

Reply:
The intention from the proposed test was not to check for correct order, although it might be possible to adapt it for such a check. It's more to check if two consumer get the same sample.